### PR TITLE
feat: Add cmd package

### PR DESCRIPTION
**Pull request by kkisic**
## Context
Execute sd-local build 

## Objective
- Add cmd package
  - root command
  - build sub command

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Comment by kkisic** [comment]
I'm thinking about how to do the test about this package

File: go.mod [hunk 1]
```diff
@@ -17,4 +17,5 @@ require (
 	gopkg.in/check.v1 v1.0.0-20190902080502-41f04d3bba15 // indirect
 	gopkg.in/yaml.v2 v2.2.8 // indirect
 	gotest.tools v2.2.0+incompatible
+	honnef.co/go/tools v0.0.1-2019.2.3
```
Comment (tk3fftk) [inline go.mod]: Is this used? and when I run `go mod tidy`,  `go.mod` is shrunk.
Do we need all of these dependencies?

```
--- a/go.mod
+++ b/go.mod
@@ -4,18 +4,11 @@ go 1.13

 require (
        github.com/go-yaml/yaml v2.1.0+incompatible
-       github.com/google/go-cmp v0.4.0 // indirect
        github.com/kr/pretty v0.2.0 // indirect
        github.com/mitchellh/go-homedir v1.1.0
-       github.com/pkg/errors v0.9.1 // indirect
-       github.com/screwdriver-cd/launcher v6.0.17+incompatible
        github.com/sirupsen/logrus v1.4.2
        github.com/spf13/cobra v0.0.5
        github.com/stretchr/testify v1.4.0
-       golang.org/x/lint v0.0.0-20200130185559-910be7a94367 // indirect
-       golang.org/x/tools v0.0.0-20200206050830-dd0d5d485177 // indirect
        gopkg.in/check.v1 v1.0.0-20190902080502-41f04d3bba15 // indirect
        gopkg.in/yaml.v2 v2.2.8 // indirect
-       gotest.tools v2.2.0+incompatible
-       honnef.co/go/tools v0.0.1-2019.2.3
 )
```

File: cmd/root.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package cmd
+
+import (
+	"github.com/spf13/cobra"
+)
+
+func newRootCmd() *cobra.Command {
+	rootCmd := &cobra.Command{
+		Use:   "sd-local",
+		Short: "Able to build in local",
```
Comment (tk3fftk) [inline cmd/root.go]: I think Able -> Enable

File: cmd/build.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+			if err != nil {
+				log.Fatal(err)
+			}
+			logger, err := buildlog.New(context.Background(), path.Join(artifactsPath, launch.LogFile), os.Stdout)
+			if err != nil {
+				log.Fatal(err)
+			}
+			go logger.Run()
+
+			// lauch.Newには API interfaceを渡すように修正したい。
```
Comment (tk3fftk) [inline cmd/build.go]: Japanese comment remains

File: cmd/build.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+)
+
+const (
+	waitIO = 1
+)
+
+func newBuildCmd() *cobra.Command {
+	buildCmd := &cobra.Command{
+		Use:   "build [job name]",
+		Short: "Run screwdriver build of the specified job name.",
```
Comment (tk3fftk) [inline cmd/build.go]: How about "Run screwdriver build" as `Short` usage?

File: cmd/build.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+			if err != nil {
+				log.Fatal(err)
+			}
+
+			api, err := screwdriver.New(config.APIURL, config.Token)
+			if err != nil {
+				log.Fatal(err)
+			}
+
+			job, err := api.Job(args[0], "./screwdriver.yaml")
```
Comment (sakka2) [inline cmd/build.go]: I think it is better to use `os.Getwd()` for using current dir same as [L53](https://github.com/screwdriver-cd/sd-local/pull/8/files#diff-ce0151979a0da3cf1233dd0fd63a5aa8R53).

File: go.mod (hunk 1 above)
Comment (sugarnaoming) [reply go.mod]: I see.
I fixed it

File: cmd/root.go (hunk 2 above)
Comment (sugarnaoming) [reply cmd/root.go]: I fixed it

File: cmd/build.go (hunk 3 above)
Comment (sugarnaoming) [reply cmd/build.go]: oh...
I fixed it

File: cmd/build.go (hunk 4 above)
Comment (sugarnaoming) [reply cmd/build.go]: I see.
I fixed it

File: cmd/build.go (hunk 5 above)
Comment (sugarnaoming) [reply cmd/build.go]: thx
I fixed it

File: cmd/build.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+			if err != nil {
+				logrus.Fatal(err)
+			}
+
+			artifactsPath := path.Join(cwd, launch.ArtifactsDir)
+			err = os.MkdirAll(artifactsPath, 0666)
+			if err != nil {
+				logrus.Fatal(err)
+			}
+			logger, err := buildlog.New(context.Background(), path.Join(artifactsPath, launch.LogFile), os.Stdout)
```
Comment (yuichi10) [inline cmd/build.go]: If we didn't use context in this function, is it better to call `context.Background()` inside of buildlog.New function?

File: cmd/build.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+			launch := launch.New(job, config, jobName, api.JWT())
+
+			err = launch.Run()
+			if err != nil {
+				logrus.Fatal(err)
+			}
+
+			// Wait for I/O processing.
+			time.Sleep(time.Second * waitIO)
```
Comment (yuichi10) [inline cmd/build.go]: Not now but someday I want to wait with I/O processing with no sleep.

File: cmd/build_test.go [hunk 8]
```diff
@@ -0,0 +1,21 @@
+package cmd
+
+import (
+	"bytes"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+type a struct {
```
Comment (yuichi10) [inline cmd/build_test.go]: can we delete this?

File: cmd/build.go (hunk 6 above)
Comment (sugarnaoming) [reply cmd/build.go]: thx
I fixed

File: cmd/build_test.go (hunk 8 above)
Comment (sugarnaoming) [reply cmd/build_test.go]: oh...
fixed

**Comment by yuichi10** [comment]
How about pass API/Config/Launch interface as argument to newBuildCmd? I think these make it possible to test cmd codes, because we can mock these API/Config/Launch things.

**Comment by kkisic** [comment]
> API/Config/Launch interface as argument to newBuildCmd

Anyway, I decided to mock API/Config/Launch without not to pass those interface as argument to newBuildCmd

File: cmd/root_test.go [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+func (mock mockAPI) JWT() string { return "" }
+
+func (mock mockLogger) Run() {}
+
+func (mock mockLogger) Stop() {}
+
+func (mock mockLaunch) Run() error { return nil }
+
+func setupTest() {
```
Comment (yuichi10) [inline cmd/root_test.go]: How about to call `setupTest` in`TestMain()`?
https://golang.org/pkg/testing/#hdr-Main

Then you do not need to call  `setupTest()` at each Test function.

Also I think just `setup()` is cleaner.

File: cmd/build_test.go [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+		root.SetArgs([]string{"test"})
+		buf := bytes.NewBuffer(nil)
+		root.SetOut(buf)
+		err := root.Execute()
+		want := ""
+		assert.Equal(t, want, buf.String())
+		assert.Nil(t, err)
+	})
+
+	t.Run("Failed build cmd when to many args", func(t *testing.T) {
```
Comment (MysticDoll) [inline cmd/build_test.go]: you mean `too`?

File: cmd/build_test.go [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+		root.SetArgs([]string{"test", "main"})
+		buf := bytes.NewBuffer(nil)
+		root.SetOut(buf)
+		err := root.Execute()
+		want := ""
+		assert.NotEqual(t, want, buf.String())
+		assert.NotNil(t, err)
+	})
+
+	t.Run("Failed build cmd when to little args", func(t *testing.T) {
```
Comment (MysticDoll) [inline cmd/build_test.go]: same as above.

File: cmd/root.go [hunk 12]
```diff
@@ -0,0 +1,22 @@
+package cmd
+
+import (
+	"github.com/spf13/cobra"
+)
+
+func newRootCmd() *cobra.Command {
+	rootCmd := &cobra.Command{
+		Use:   "sd-local",
+		Short: "Enable to build in local",
```
Comment (s-yoshika) [inline cmd/root.go]: How about the message like `Run build in local`. I think `Enable to~` is not suitable for the description of the command.

Comment (sugarnaoming) [reply cmd/root.go]: I fixed it

File: cmd/build_test.go (hunk 10 above)
Comment (sugarnaoming) [reply cmd/build_test.go]: I fixed it

File: cmd/build_test.go (hunk 11 above)
Comment (sugarnaoming) [reply cmd/build_test.go]: I fixed it

File: cmd/root_test.go (hunk 9 above)
Comment (sugarnaoming) [reply cmd/root_test.go]: I fixed it

**Pull request merged**